File: riveredge-backend/src/core/schemas/approval_instance.py
"""
审批实例 Schema 模块

定义审批实例相关的 Pydantic Schema，用于数据验证和序列化。
"""

from pydantic import BaseModel, Field, ConfigDict, field_validator
from typing import Optional, Dict, Any
from datetime import datetime
from uuid import UUID


class ApprovalInstanceBase(BaseModel):
    """审批实例基础 Schema"""
    process_uuid: UUID = Field(..., description="关联流程UUID")
    title: str = Field(..., max_length=200, description="审批标题")
    content: Optional[str] = Field(None, description="审批内容")
    data: Optional[Dict[str, Any]] = Field(None, description="审批数据")


class ApprovalInstanceCreate(ApprovalInstanceBase):
    """创建审批实例 Schema"""
    pass


class ApprovalInstanceUpdate(BaseModel):
    """更新审批实例 Schema"""
    title: Optional[str] = Field(None, max_length=200, description="审批标题")
    content: Optional[str] = Field(None, description="审批内容")
    data: Optional[Dict[str, Any]] = Field(None, description="审批数据")
    status: Optional[str] = Field(None, max_length=20, description="审批状态")
    current_node: Optional[str] = Field(None, max_length=100, description="当前节点")
    current_approver_id: Optional[int] = Field(None, description="当前审批人ID")
    
    @field_validator('status')
    @classmethod
    def validate_status(cls, v):
        """验证审批状态"""
        if v is None:
            return v
        allowed_statuses = ['pending', 'approved', 'rejected', 'cancelled']
        if v not in allowed_statuses:
            raise ValueError(f'审批状态必须是 {allowed_statuses} 之一')
        return v


class ApprovalInstanceAction(BaseModel):
    """审批操作 Schema"""
    action: str = Field(..., description="操作类型（approve、reject、cancel、transfer）")
    comment: Optional[str] = Field(None, description="审批意见")
    transfer_to_user_id: Optional[int] = Field(None, description="转交目标用户ID（仅转交时使用）")
    
    @field_validator('action')
    @classmethod
    def validate_action(cls, v):
        """验证操作类型"""
        allowed_actions = ['approve', 'reject', 'cancel', 'transfer']
        if v not in allowed_actions:
            raise ValueError(f'操作类型必须是 {allowed_actions} 之一')
        return v


class ApprovalInstanceResponse(ApprovalInstanceBase):
    """审批实例响应 Schema"""
    uuid: UUID = Field(..., description="审批实例UUID")
    tenant_id: int = Field(..., description="组织ID")
    process_uuid: UUID = Field(..., description="关联流程UUID")
    status: str = Field(..., description="审批状态")
    current_node: Optional[str] = Field(None, description="当前节点")
    current_approver_id: Optional[int] = Field(None, description="当前审批人ID")
    inngest_run_id: Optional[str] = Field(None, description="Inngest 运行ID")
    submitter_id: int = Field(..., description="提交人ID")
    submitted_at: datetime = Field(..., description="提交时间")
    completed_at: Optional[datetime] = Field(None, description="完成时间")
    created_at: datetime = Field(..., description="创建时间")
    updated_at: datetime = Field(..., description="更新时间")
    
    model_config = ConfigDict(from_attributes=True)

File: riveredge-frontend/src/services/approvalInstance.ts
/**
 * 审批实例管理服务
 * 
 * 提供审批实例的 CRUD 操作和审批操作功能。
 * 注意：所有 API 自动过滤当前组织的审批实例
 */

import { apiRequest } from './api';

export interface ApprovalInstance {
  uuid: string;
  tenant_id: number;
  process_uuid: string;
  title: string;
  content?: string;
  data?: Record<string, any>;
  status: 'pending' | 'approved' | 'rejected' | 'cancelled';
  current_node?: string;
  current_approver_id?: number;
  inngest_run_id?: string;
  submitter_id: number;
  submitted_at: string;
  completed_at?: string;
  created_at: string;
  updated_at: string;
}

export interface ApprovalInstanceListParams {
  skip?: number;
  limit?: number;
  status?: string;
  submitter_id?: number;
  current_approver_id?: number;
}

export interface CreateApprovalInstanceData {
  process_uuid: string;
  title: string;
  content?: string;
  data?: Record<string, any>;
}

export interface UpdateApprovalInstanceData {
  title?: string;
  content?: string;
  data?: Record<string, any>;
  status?: string;
  current_node?: string;
  current_approver_id?: number;
}

export interface ApprovalInstanceActionData {
  action: 'approve' | 'reject' | 'cancel' | 'transfer';
  comment?: string;
  transfer_to_user_id?: number;
}

/**
 * 按 entity 获取审批状态（统一入口，供 UniApprovalPanel 等使用）
 */
export interface ApprovalProcessGraph {
  nodes?: Array<{
    id: string;
    type?: string;
    position?: { x: number; y: number };
    data?: Record<string, unknown>;
  }>;
  edges?: Array<{ id?: string; source: string; target: string }>;
}

export interface ApprovalProcessBrief {
  uuid: string;
  name: string;
  code: string;
  nodes: ApprovalProcessGraph;
}

export interface ApprovalInstanceBrief {
  uuid: string;
  submitter_id: number;
  submitter_name: string;
  submitted_at?: string | null;
  title?: string;
}

export interface ApprovalNodeExecution {
  action?: string;
  action_label?: string;
  action_by?: number;
  action_by_name?: string;
  action_at?: string | null;
  comment?: string | null;
  from_node?: string | null;
  to_node?: string | null;
  source?: string;
  field_changes?: Array<{ field?: string; label?: string; from?: string; to?: string }>;
}

export interface ApprovalNodeOverview {
  node_id: string;
  label: string;
  node_type: string;
  is_current: boolean;
  status: 'pending' | 'approved' | 'rejected' | 'waiting' | 'skipped' | string;
  eligible_approvers: Array<{ user_id: number; name: string }>;
  executions: ApprovalNodeExecution[];
}

export interface ApprovalStatusResponse {
  has_flow: boolean;
  has_instance?: boolean;
  status?: string;
  current_node?: string;
  current_approver_id?: number;
  process?: ApprovalProcessBrief | null;
  instance?: ApprovalInstanceBrief | null;
  nodes_overview?: ApprovalNodeOverview[];
  can_edit_during_approval?: boolean;
  editable_fields?: string[] | '*';
  node_capabilities?: {
    allow_transfer?: boolean;
    allow_add_sign?: boolean;
    allow_edit_during_approval?: boolean;
  };
  tasks?: Array<{
    uuid: string;
    node_id?: string;
    approver_id?: number;
    status?: string;
    action_at?: string;
    comment?: string;
  }>;
  history?: Array<{
    action?: string;
    action_by?: number;
    action_by_name?: string;
    action_at?: string;
    comment?: string;
    from_node?: string;
    to_node?: string;
  }>;
}

export async function getApprovalStatus(
  entityType: string,
  entityId: number,
): Promise<ApprovalStatusResponse> {
  return apiRequest<ApprovalStatusResponse>('/core/approval-instances/status', {
    params: { entity_type: entityType, entity_id: entityId },
  });
}

/**
 * 获取审批实例列表
 */
export async function getApprovalInstanceList(params?: ApprovalInstanceListParams): Promise<ApprovalInstance[]> {
  return apiRequest<ApprovalInstance[]>('/core/approval-instances', {
    params,
  });
}

/**
 * 获取审批实例详情
 */
export async function getApprovalInstanceByUuid(approvalInstanceUuid: string): Promise<ApprovalInstance> {
  return apiRequest<ApprovalInstance>(`/core/approval-instances/${approvalInstanceUuid}`);
}

/**
 * 创建审批实例（提交审批）
 */
export async function createApprovalInstance(data: CreateApprovalInstanceData): Promise<ApprovalInstance> {
  return apiRequest<ApprovalInstance>('/core/approval-instances', {
    method: 'POST',
    data,
  });
}

/**
 * 更新审批实例
 */
export async function updateApprovalInstance(approvalInstanceUuid: string, data: UpdateApprovalInstanceData): Promise<ApprovalInstance> {
  return apiRequest<ApprovalInstance>(`/core/approval-instances/${approvalInstanceUuid}`, {
    method: 'PUT',
    data,
  });
}

/**
 * 删除审批实例
 */
export async function deleteApprovalInstance(approvalInstanceUuid: string): Promise<void> {
  return apiRequest<void>(`/core/approval-instances/${approvalInstanceUuid}`, {
    method: 'DELETE',
  });
}

/**
 * 执行审批操作（同意、拒绝、取消、转交）
 */
export async function performApprovalAction(approvalInstanceUuid: string, action: ApprovalInstanceActionData): Promise<ApprovalInstance> {
  return apiRequest<ApprovalInstance>(`/core/approval-instances/${approvalInstanceUuid}/action`, {
    method: 'POST',
    data: action,
  });
}

export interface ApprovalInstanceBatchActionData {
  instance_uuids: string[];
  action: 'approve' | 'reject';
  comment?: string;
}

export interface ApprovalInstanceBatchResult {
  success_count: number;
  failure_count: number;
  failures: Array<{ uuid: string; error: string }>;
}

export async function batchPerformApprovalActions(
  data: ApprovalInstanceBatchActionData,
): Promise<ApprovalInstanceBatchResult> {
  return apiRequest<ApprovalInstanceBatchResult>('/core/approval-instances/batch-action', {
    method: 'POST',
    data,
  });
}

